auction/chaincode-go/smart-contract/auction.go

```diff
@@ -0,0 +1,540 @@
+
+	err = ctx.GetStub().PutState(auctionID, closedAuction)
+	if err != nil {
+		return fmt.Errorf("failed to close auction: %v", err)
+	}
+
+	return nil
+}
+
+// EndAuction both changes the auction status to closed and calculates the winners
+// of the auction
+func (s *SmartContract) EndAuction(ctx contractapi.TransactionContextInterface, auctionID string) error {
+
+	auctionBytes, err := ctx.GetStub().GetState(auctionID)
+	if err != nil {
+		return fmt.Errorf("failed to get auction %v: %v", auctionID, err)
+	}
+
+	if auctionBytes == nil {
+		return fmt.Errorf("Auction interest object %v not found", auctionID)
+	}
+
+	var auctionJSON Auction
+	err = json.Unmarshal(auctionBytes, &auctionJSON)
+	if err != nil {
+		return fmt.Errorf("failed to create interst object JSON: %v", err)
+	}
+
+	// Check that the auction is being ended by the seller
+
+	// get ID of submitting client
+	clientID, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return fmt.Errorf("failed to get client identity %v", err)
+	}
+
+	Seller := auctionJSON.Seller
+	if Seller != clientID {
+		return fmt.Errorf("auction can only be ended by seller: %v", err)
+	}
+
+	Status := auctionJSON.Status
+	if Status != "closed" {
+		return fmt.Errorf("Can only end a closed auction")
+	}
+
+	// get the list of revealed bids
+
+	revealedBidMap := auctionJSON.RevealedBids
+	if len(auctionJSON.RevealedBids) == 0 {
+		return fmt.Errorf("No bids have been revealed, cannot end auction: %v", err)
+	}
+
+	// sort the map of revealed bids to make it easier to calculate winners
+	// if bids are tied, fill smaller bids first
+	var Bidders []FullBid
+
+	for _, bid := range revealedBidMap {
+		Bidders = append(Bidders, bid)
+	}
+
+	sort.Slice(Bidders, func(p, q int) bool {
+		if Bidders[p].Price > Bidders[q].Price {
+			return true
+		}
+		if Bidders[p].Price < Bidders[q].Price {
+			return false
+		}
+		return Bidders[p].Quantity < Bidders[q].Quantity
+	})
+
+	i := 0
+	remainingQuantity := auctionJSON.Quantity
+
+	// calculate the winners
+	for remainingQuantity > 0 {
+
+		// create the next winning bid
+		winner := Winners{
+			Buyer:    Bidders[i].Buyer,
+			Quantity: Bidders[i].Quantity,
+		}
+
+		// add them to the list of winners and change the winning price
+		auctionJSON.Winners = append(auctionJSON.Winners, winner)
+		auctionJSON.Price = Bidders[i].Price
+
+		// Calculate the quantity that goes to the winner
+		// if there is sufficient quantity to give them the full bid
+		if remainingQuantity > Bidders[i].Quantity {
+			remainingQuantity = remainingQuantity - Bidders[i].Quantity
+
+			// if there is not, give the remainder
+		} else {
+			auctionJSON.Winners[i].Quantity = remainingQuantity
+			remainingQuantity = 0
+		}
+		i++
+	}
+
+	// check if there is a winning bid that has yet to be revealed
+	err = queryAllBids(ctx, auctionJSON.Price, auctionJSON.RevealedBids, auctionJSON.PrivateBids)
+	if err != nil {
+		return fmt.Errorf("Cannot close auction: %v", err)
+	}
+
```

Comment:
This ought to be done before calculating the winning bids since it may make that calculation pointless.